ros_industrial_cmake_boilerplate/cmake/cmake_tools.cmake

```diff
@@ -219,6 +219,41 @@ macro(install_pkgxml)
   install(FILES package.xml DESTINATION share/${PROJECT_NAME})
 endmacro()
 
+# Create a default *-config.cmake.in with simple dependency finding
+function(make_default_package_config)
+    set(options HAS_TARGETS)
+    set(multiValueArgs DEPENDENCIES)
+    cmake_parse_arguments(ARG "${options}" "" "${multiValueArgs}" ${ARGN})
+
+    string(CONCAT ricb_pkgconfig
+        "#########################################################################\n"
+        "# Default *-config.cmake file created by ros-industrial-cmake-boilerplate\n"
+        "#########################################################################\n\n"
+        "@PACKAGE_INIT@\n\n"
+        "set(@PROJECT_NAME@_FOUND ON)\n"
+        "set_and_check(@PROJECT_NAME@_INCLUDE_DIRS \"\${PACKAGE_PREFIX_DIR}/include\")\n"
+        "set_and_check(@PROJECT_NAME@_LIBRARY_DIRS \"\${PACKAGE_PREFIX_DIR}/lib\")\n"
+    )
+
+    if (ARG_DEPENDENCIES)
+        set(find_dep_cmds "include(CMakeFindDependencyMacro)")
+        foreach(dep IN LISTS ARG_DEPENDENCIES)
+            list(APPEND find_dep_cmds "find_dependency(${dep})")
```

Comment:
To enable support of components I would do something along the lines to what is shown below.

The  components are stored in `<PackageName>_FIND_COMPONENTS` when you call find_package in your project. 

``` cmake
if (${${dep}_FIND_COMPONENTS})
  list(APPEND find_dep_cmds "find_dependency(${dep} COMPONENTS ${${dep}_FIND_COMPONENTS})")
else()
  list(APPEND find_dep_cmds "find_dependency(${dep})")
endif()
```

Reply:
Also when need to add a check like the one below when adding COMPONENTS.
``` cmake
if(${CMAKE_VERSION} VERSION_LESS "3.15.0")
    find_package(Boost REQUIRED COMPONENTS system)
else()
    find_dependency(Boost COMPONENTS system)
endif()
```

Reply:
I'll see if I can play around with it and try testing by tomorrow. Two followup questions:
 - Is there any current ricb package that provides different components? I think I've only seen it for Boost and PCL.
 - I can't find an explanation for the different version behavior of `find_package` vs `find_dependency`. What motivates the check?

Reply:
Ignition is another one. We also plan to update several tesseract packages to provide components like tesseract_collision, tesseract_kinematics, etc.

I not sure where I got `3.15.0`, but did some digging in the cmake [release notes](https://github.com/Kitware/CMake/blob/859241d2bbaae83f06c44bc21ab0dbd38725a3fc/Help/release/3.9.rst) and it was actually resolved in version `3.9.0`. It was not forwarding extract arguments to the find_package call which is why COMPONENTS were not supported prior to 3.9.0.

Reply:
I just want to make sure I fully understand everything here. There seem to be two cases to me:

1. A downstream package wants to call `find_package(tesseract_kinematics COMPONENTS kdl)`, for example, and so `tesseract_kinematics-config.cmake` must check the `*_FIND_COMPONENTS` variable and include `find_dependency(kdl)`.
2. A RICB package (say, `tesseract_ignition`) depends on a particular component (`ignition-rendering::ogre1`, e.g.) and so the generated `tesseract_ignition-config.cmake` file must contain `find_dependency(ignition-rendering COMPONENTS ogre1)` (with version check)

Am I correct in my understanding that the goal is support use case 1 here and not 2? And in the case of 2, a custom template is still needed rather than the autogenerated config file?

Reply:
The second case is what is need, because the first would always fail if it was not an actual component. If you call find_package with components it actually looks for the following cmake config file. `tesseract_kinematics-kdl-config.cmake` and the other case would be `ignition-rendering-oger1-config.cmake`. 

1.
``` cmake 
find_package(tesseract_kinematics COMPONENTS kdl) -> find_dependency(tesseract_kinematics COMPONENTS kdl) 
```
2.
``` cmake
find_package(ignition-rendering COMPONENTS ogre1) -> find_dependency(ignition-rendering COMPONENTS ogre1) 
```

Reply:
I'm confused then because the `_FIND_COMPONENTS` variable seems to be only for the first case, where the RICB package is providing components for downstream packages. Specifically, calling `find_package(Boost COMPONENTS system)` in `CMakeLists.txt` of `somepkg`, for example, does not define a `Boost_FIND_COMPONENTS` within the CMake context of `somepkg` which could be used when autogenerating the `-config.cmake` file. Am I not understanding this problem correctly?

Reply:
It looks like this is not generated when I thought. This is generated to be used in the packages config.cmake so you can tell if someone called find package on your package with components, so this is not going to work. It looks like using component does not do anything special it just so you do not have to call multiple find packages.

Doing this 

``` cmake
find_package(Boost COMPONENTS system graph)
```

is the same as the following it just does it behind the scene

``` cmake
find_package(Boost-system)
find_package(Boost-graph)
```

As a result we can just do the following with what you have now but need to document it.

``` cmake
find_package(Boost COMPONENTS REQUIRED system graph)

configure_package(
  DEFAULT_TEMPLATE
  NAMESPACE opw_kinematics
  TARGETS ${PROJECT_NAME}
  DEPENDENCIES 
    Boost-system
    Boost-graph
)
```

Reply:
Is this a general rule or just a Boost convention? The CMake docs seem to say packages are completely free to interpret components however they want.

We could also do something like providing components as separate lists, like "Boost;system;graph" though that seems a bit confusing.

Reply:
Your correct. I believe this is the stand approach, but you are correct it is up to the package config on how they are handled. If it is wrapped in quotes would it be treated as one entry in the list?

``` cmake
configure_package(
  DEFAULT_TEMPLATE
  NAMESPACE opw_kinematics
  TARGETS ${PROJECT_NAME}
  DEPENDENCIES 
    "Boost COMPONENTS system graph"
    tesseract_scene_graph
)
```

Reply:
Yes, the space separated  quoted lined will be a single item that can be processed (not the semi-colon one I suggested before). I'll spend a few minutes to see if it can easily be handled.

Reply:
So yes, the example above with the quotes works automatically. If you're ok with skipping the version check (3.9 is five years old now) no further changes needs to be made.

Reply:
Yea that works for me.